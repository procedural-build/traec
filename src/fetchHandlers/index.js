/**
 * fetchHandlers are low-level functions that simply return:
 *
 * 1. the fetchParam (fetch parameters) which is an object (dictionary) that just provides:
 *    * url: The URL of the endpoint to fetch
 *    * method: The HTTP method to
 *    * apiId: The unique_id of the API which is used to generate the "handlerMap" on app initialization (see more below)
 * 2. the stateParams which is an object containing a set of parameters that determine how to store/unpack the fetched state into the Redux state.  The stateParams primarily contains:
 *    * stateSetFunc: A function that takes the fetch response (action.payload) and unpacks it into the Redux Store (state)
 *    * stateCheckFunc: A function that is used to check if the data already exists in the Redux store (returns true or false)
 *
 * NOTE 1: *handlerMap* - On app initialization all of the fetchHandlers that have an apiId are mapped into an object that allows
 * one to reference the fetchHandler by apiId (which may be obtained from Swagger documentation) and HTTP method.  The handlerMap
 * maps to an other plain JS object which contains:
 * * **fetchHandler**: the actual fetchHandler function
 * * **requiredParams**: required Parameters that must be in the components props
 * * **queryParams**: optional Parameters that may be provided
 *
 * This allows you to get the appropriate fetchHandler function by API reference ID and NOT having to remember the name of all
 * individual fetchHandler functions for each API endpoint.  This allows you to use the appropriate fetchHandler within a React
 * component as in the example below.
 *
 * NOTE 2: *fetchBindings* - "Binding" functions such as "fetchAndFlagState" are functions which will "bind" onto the class and
 * do the following BEFORE calling the fetchHandler:
 * * Check that all of the parameters required for the fetch (ie. projectId, refId, trackerId, etc.) can be obtained from the component props (this.props).  If a parameter is missing then the fetchBinding returns with no action.  This means that the prop may still be loading other data that is required in order to determine the missing parameter and this fetch may be called in a later "componentDidUpdate" call.
 * * Check that the data does not already exist in the Redux state (by calling the stateCheckFunc).  If the stateCheckFunc returns "true" then the data has already been previously fetched and the fetch is blocked UNLESS the argument "ignoreStateCheck" is set to true.
 * * If the above two checks pass then the fetchHandler is called in order to retrieve the data.  At the time of calling then  the components state is flagged that that fetch handler URL has been called, so we can safely include a call to this binding in all component update methods in order to continuously check and re-fetch data as required to robustly obtain all data required to render a particular component.
 *
 * @namespace fetchHandlers
 * @example
 * import Traec from 'traec'
 *
 * class MyComponent extends React.Component {
 *     constructor(props) {
 *         super(props)
 *
 *         this.requiredFetches = [
 *             new Traec.Fetch('project_tracker', 'list'),
 *             new Traec.Fetch('tracker', 'list'),
 *         ]
 *     }
 *
 *     componentDidMount() {
 *         Traec.fetchRequired.bind(this)()
 *     }
 *
 *     componentDidUpdate() {
 *         Traec.fetchRequired.bind(this)()
 *     }
 * }
 *
 */

// AUTH
export * from "./auth/registration";

// COMMIT
export * from "./commit/commitEdges";
export * from "./commit/commitBranch";
export * from "./commit/commit";

//COMPANY
export * from "./company/company";
export * from "./company/companyAuthGroup";
export * from "./company/companyInvite";
export * from "./company/companyMembers";
export * from "./company/companyReporting";
export * from "./company/companyReportInputs";
export * from "./company/companyEmails";
export * from "./company/dispatch";
export * from "./company/companyTargets";
export * from "./company/companyIndicator";

//PROJECT
export * from "./project/projectReporting";
export * from "./project/projectDiscipline";
export * from "./project/projectMember";
export * from "./project/projectInvite";
export * from "./project/project";
export * from "./project/projectSetup";
export * from "./project/projectAuthGroup";
export * from "./project/projectInputs";
export * from "./project/projectSupplierRequest";
export * from "./project/projectEmails";
export * from "./project/projectIndicators";
export * from "./project/projectStatus";

//TENANT
export * from "./tenant";

//UTILS
export * from "./nodes";
export * from "./target";
export * from "./tree";
export * from "./head";
export * from "./tracker";
export * from "./treeDescription";
export * from "./document";
export * from "./ref";
export * from "./metric";
export * from "./convFactors";
export * from "./indicators";
export * from "./comment";
export * from "./store";
export * from "./mailReceiver";
