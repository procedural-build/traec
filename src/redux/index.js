import store from "./store";

/**
 * Redux related objects, store and utility functions
 * @namespace redux
 */

export default {
  store,
};
