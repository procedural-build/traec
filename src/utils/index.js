/**
 * Redux Middleware for API Fetches
 * @namespace utils
 */

export { camelCaseToSentence } from "./text";
export * from "./jsobject";
export * from "./uuidUtils";
