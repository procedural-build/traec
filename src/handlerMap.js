import * as defaultFetchHandlers from "./fetchHandlers";

/**
 * Object for getting the fetch handler functions to use based on API endpoint.
 * The map also includes methods for checking if the object already exists in
 * the Redux State
 *
 * Note that this method is called on initialization of the 'traec' library
 * in order to provide a mapping to all of the functions in fetchHandlers.
 *
 * Generally users will not need to call this function directly.  Use the
 * fetchManager "Fetch" class instead.
 *
 * The FIRST key is the API reference (ie api_company_list) but with the
 * prefix "api_" removed and the action-suffix (ie. _list) also removed.  The
 * second KEY is the http method which are:
 * * list    = list
 * * read    = read
 * * post    = create
 * * put     = update
 * * patch   = update
 * * delete  = delete
 *
 * @method
 * @memberof utils
 */
export const makeHandlerMap = function(fetchHandlers = null) {
  console.log("MAKING fetchHandler MAP");
  let handlerMap = {};
  if (fetchHandlers == null) {
    fetchHandlers = defaultFetchHandlers;
  }
  for (let [funcName, fetchHandler] of Object.entries(fetchHandlers)) {
    try {
      let { fetchParams, stateParams } = fetchHandler({});
      if (!fetchParams) {
        console.warn("Skipping function in fetchhandler with fetchParams undefined", fetchHandler);
        continue;
      }
      let { apiId, method } = fetchParams;
      if (apiId) {
        //console.log(`FOUND API ID ${apiId}`)
        let parts = apiId.split("_");
        let endIndex = apiId.endsWith("partial_update") ? parts.length - 2 : parts.length - 1;
        let prefix = parts.slice(1, endIndex).join("_");
        // Get the action of the fetch
        let action = parts.slice(endIndex, parts.length).join("_");
        if (["POST", "PUT", "PATCH"].indexOf(method) > -1) {
          action = method.toLowerCase();
        }
        let actionMap = handlerMap[prefix] || {};
        actionMap[action] = {
          fetchHandler: fetchHandler,
          requiredParams: fetchParams.requiredParams || [],
          queryParams: fetchParams.queryParams || []
        };
        // Assign the updated action map to the handlerMap
        Object.assign(handlerMap, { [prefix]: actionMap });
      }
    } catch (err) {
      console.warn("Error constructing handlerMap", funcName, err);
      continue;
    }
  }
  console.log(" - generated handlerMap", handlerMap);
  return handlerMap;
};

export const handlerMap = makeHandlerMap();
