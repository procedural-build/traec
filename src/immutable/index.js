import Im from "immutable";
import * as utils from "./utils/stateBindins";

/**
 * Patched Immutable.js library with some common methods that are used
 * in the API calls and fetchHandlers specific to Traec.
 *
 * If using immutable then import from here instead of immutable directly,
 * like so:
 * @example
 * // Change from this:
 * import Im from 'immutable';
 * //to this:
 * import Im from 'traec/immutable'
 * //or this:
 * import Traec from 'traec'
 * const {Im} = Traec
 *
 * @namespace immutable
 */

//Check if the browser is IE, and display the error message if true
var isIE = false || !!document.documentMode;

if (isIE) {
  alert(
    "The Sustainability Tool is no longer supported on Internet Explorer. Please use another internet browser application or contact us for more information. "
  );
  window.location = "error.html";
}

// Patch all of the functions from utils onto the Map prototype
// to make them available within Immutable.map objects
// NOTE: The Redux state should therefore be a Map object
for (let [name, func] of Object.entries(utils)) {
  Im.Map.prototype[name] = func;
}

export default Im;
