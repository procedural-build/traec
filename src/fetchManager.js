import { fetchToState, toggleForm } from "./redux/actionCreators";
import { handlerMap } from "./handlerMap";
import { hasFetched } from "./redux/fetchCache";
import store from "./redux/store";

/**
 * This is a wrapper around the fetchHandlers which manages the setting, getting
 * and dispatching of the fetchHandler methods.  So instead of calling fetchHandlers
 * from within a connected React component like as per example below.
 *
 * This class includes "hooks" which may be used to trigger actions and other events.
 * Hooks available are (see functions for more information):
 * * preUpdateHook
 * * preDispatchHook
 *
 * @param {string} apiId The apiId of the fetchHandler that this relates to (obtained from Swagger API docs) with leading api_ and trailing action (_read, _update, etc.) stripped
 * @param {string} method The fetchHandler HTTP method that will be used to call the API (ie. read, list, put, patch, delete)
 * @param {object} params the parameters object {fetchParams, stateParams}.
 * @param {object} overrides overrides to any key of this class object (useful for defining hooks at instantiation)
 * @return {object} an instance of this class
 *
 * @example
 * import Traec from 'traec'
 *
 * class MyComponent extends React.Component {
 *     constructor(props) {
 *         super(props)
 *
 *         this.requiredFetches = [
 *             new Traec.Fetch('project_tracker', 'list'),
 *             new Traec.Fetch('tracker', 'list'),
 *         ]
 *     }
 *
 *     componentDidMount() {
 *         Traec.fetchRequired.bind(this)()
 *     }
 *
 *     componentDidUpdate() {
 *         Traec.fetchRequired.bind(this)()
 *     }
 * }
 */
export default class Fetch {
  constructor(apiId, method, params, overrides = {}) {
    /* HOOKS 
        Override these to do something prior to dispatch
        (ie. validate action.fetchParms.body or display a notification)*/
    this.preDispatchHook = action => action;
    this.preUpdateHook = args => args;

    // Properties of this can be overrided on instantiation
    for (let key of Object.keys(overrides)) {
      this[key] = overrides[key];
    }

    // Store the original apiId and method calls
    this._apiId = apiId;
    this._method = method;

    // Get the fetchHandler from the map created on app initialization
    this.fetchHandler = handlerMap[apiId][method].fetchHandler;

    // Set a default cacheTimeout
    this.defaultCacheTimeout = 3600;

    // Initialize some fetch Parameters
    this.update(params);
  }

  get params() {
    return {
      fetchParams: this._fetchParams,
      stateParams: this._stateParams
    };
  }

  get url() {
    return this._fetchParams ? this._fetchParams.url : null;
  }

  get requiredParams() {
    return (this._fetchParams ? this._fetchParams.requiredParams : []) || [];
  }

  get queryParams() {
    return (this._fetchParams ? this._fetchParams.queryParams : {}) || {};
  }

  get cacheTimout() {
    return this._fetchParams.cacheTime || this.defaultCacheTimeout * 1000;
  }

  get method() {
    return this._fetchParams ? this._fetchParams.method : null;
  }

  /**
   * Check if this fetch has been made within the cache timeout period
   * @return {boolean} True if we have requested data from this endpoint within cacheTimeout period
   */
  hasFetched() {
    // Check the Redux fetch
    if (this.method == "GET") {
      return hasFetched(store.getState(), this._fetchParams, this.cacheTimout);
    }
    return false;
  }

  /**
   * If there is a stateCheckFunc in the stateParams then check if data already exists in the redux store
   * @return {boolean} or null if stateCheckFunc does not exist.
   */
  hasData() {
    // If there is a stateCheckFunc in the stateParams then use that to check
    // if data already exists in the redux store.
    let { stateCheckFunc } = this.params.stateParams;
    if (!stateCheckFunc) {
      return null;
    }
    return stateCheckFunc(store.getState());
  }

  /**
   * Update the parameters for this Fetch based on
   *
   * **Notes**:
   * * the second argument is redundant (you can pass the queryParams through the first "params" parameter if you want)
   * * this function first calls "preUpdateHook" which provides an opportunity to modify the parameters that are passed to the fetchHandler.  See the documentation for "preUpdateHook"
   *
   * @param {object} params The parameters in an object that will be destructured in the fetchHandler
   * @param {object} [queryParams=] The queryParameters in an object that will be destructured in the fetchHandler
   * @return {object} the parameters object {fetchParams, stateParams}.
   */
  update(params = {}, queryParams = {}) {
    params = this.preUpdateHook({ ...params, ...queryParams });
    let p = this.fetchHandler(params);
    this._fetchParams = p.fetchParams;
    this._stateParams = p.stateParams;
    return params;
  }

  /**
   * Update/override the stateParams of this Fetch
   * @param {object} [obj={}] The fetch parameters that should be overriden
   * @return {object} the parameters object {fetchParams, stateParams}.
   */
  updateStateParams(obj = {}) {
    Object.assign(this._stateParams, obj);
    return this.params;
  }

  /**
   * Update/override the fetchParams of this Fetch
   * @param {object} [obj={}] The fetch parameters that should be overriden
   * @return {object} the parameters object {fetchParams, stateParams}.
   */
  updateFetchParams(obj = {}) {
    Object.assign(this._fetchParams, obj);
    return this.params;
  }

  /**
   * If the stateParams has a "formVisPath" field then that will be toggled to
   * show/hide the form visibility parameter in Redux.
   *
   * **Note**:
   * * it is expected that there are commponents connected to Redux and watching the formVisPath to determine if the required form for this fetch should be displayed. This is done automatically by the BaseForm class of 'traec-react'.
   */
  toggleForm() {
    store.dispatch(toggleForm(this.params.stateParams));
  }

  /**
   * Dispatch the fetch (ie. request from the API) IFF the request has not already been made within cacheTimeout
   * period before the current request.
   *
   * **Note**:
   * * this calls "preDispatchHook" which provides an opportunity to trigger an action just prior to the actual fetch-action being dispatched.  This is useful for displaying an alert or popup confirmation and/or triggering a popup that may wait for a success or failure.
   */
  dispatch() {
    if (this.hasFetched()) {
      return null;
    }
    let action = fetchToState(this.params);
    this.preDispatchHook(action);
    store.dispatch(action);
  }

  /**
   * Check if the object has the required fields that can be destructured for the fetchHandler
   * @param {object} [obj={}] The fetch parameters that should be overriden
   * @return {boolean}
   */
  hasRequiredParams(obj) {
    for (let paramName of this.requiredParams) {
      if (!obj[paramName]) {
        //console.warn("Required parameters not available")
        return false;
      }
    }
    return true;
  }

  // Setter functions for
  dispatchFromProps(props, fetchedUrls, stateSet, forceFetch = false) {
    // Call the fetchHandler with the arguments provided
    // NOTE: The newProps returned may be modified by a preUpdateHook
    let newProps = this.update(props);
    // Check that the required parameters are met
    if (!this.hasRequiredParams(newProps)) {
      return null;
    }
    // Check if the URL to be fetched has already been requested in the state
    let { fetchParams } = this.params;
    fetchedUrls = fetchedUrls || {};
    if (!fetchedUrls[fetchParams.url] || forceFetch == true) {
      // Dispatch this action
      this.dispatch();
      // Update the fetched Urls flag with the flagSetter
      Object.assign(fetchedUrls, { [fetchParams.url]: true });
      stateSet({ fetchedUrls });
    }
  }
}
