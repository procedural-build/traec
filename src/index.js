import * as fetchHandlers from "./fetchHandlers";
import * as fetchBindings from "./fetchBindings";
import { handlerMap, makeHandlerMap } from "./handlerMap";
import { fetchRequired } from "./fetchBindings/utils";
import Fetch from "./fetchManager";
import Im from "./immutable";
import Redux from "./redux";
import * as utils from "./utils";

export * from "./handlerMap";
export * from "./fetchBindings/utils";

export default {
  fetchHandlers,
  fetchBindings,
  handlerMap,
  makeHandlerMap,
  fetchRequired,
  Fetch,
  Im,
  Redux,
  utils,
};
