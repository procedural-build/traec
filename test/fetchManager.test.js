import { handlerMap } from "../src";

describe("fethcManager", () => {
  it("should contain a test", () => {});
});
