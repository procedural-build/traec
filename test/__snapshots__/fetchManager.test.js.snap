// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fethcManager should contain a test 1`] = `undefined`;
